garage/tf/models/base.py

```diff
@@ -0,0 +1,297 @@
+
+
+class Model(BaseModel):
+    r"""
+    Model class for TensorFlow.
+
+    A TfModel only contains the structure/configuration of the underlying
+    computation graphs. Connectivity information are all in Network class.
+    A TfModel contains zero or more Network.
+
+    When a Network is created, it reuses the parameter from the
+    model and can be accessed by calling model.networks['network_name'],
+    If a Network is built without given a name, the name "default" will
+    be used.
+
+    ***
+    Do not call tf.global_variable_initializers() after building a model as it
+    will reassign random weights to the model.
+    The parameters inside a model will be initialized when calling build().
+    ***
+
+    Pickling is handled automatcailly. The target weights should be assigned to
+    self._default_parameters before pickling, so that the newly created model
+    can check if target weights exist or not. When unpickled, the unserialized
+    model will load the weights from self._default_parameters.
+
+    The design is illustrated as the following:
+
+         input_1                      input_2
+            |                            |
+    ============== Model (TfModel)===================
+    |       |                            |          |
+    |       |            Parameters      |          |
+    |    =============  /           \  ============ |
+    |    |  default  | /             \ | Network2 | |
+    |    | (Network) |/               \|(Network) | |
+    |    =============                 ============ |
+    |       |                            |          |
+    =================================================
+            |                            |
+            |                            |
+    (model.networks['default'].outputs)  |
+                        model.networks['Network2'].outputs
+
+
+    Examples are also available in tests/garage/tf/models/test_model.
+
+    Args:
+      name (str): Name of the model. It will also become the variable scope
+      of the model. Every model should have a unique name.
+    """
+
+    def __init__(self, name):
+        super().__init__()
+        self._name = name or type(self).__name__  # name default to class
+        self._networks = {}
+        self._default_parameters = None
+
+    def build(self, *inputs, name=None):
+        """
+        Build a Network with the given input(s).
+
+        ***
+        Do not call tf.global_variable_initializers() after building a model
+        as it will reassign random weights to the model.
+        The parameters inside a model will be initialized when calling build().
+        ***
+
+        It uses the same, fixed variable scope for all Networks, to ensure
+        parameter sharing. Different Networks must have an unique name.
+
+        Args:
+          inputs: Tensor input(s), recommended to be position arguments, e.g.
+            def build(self, state_input=None, action_input=None, name=None).
```

Comment:
your example uses keyword arguments (e.g. `build(self, state_input=None, action_input=None, name=None)`), not positional arguments (e.g. `build(self, state_input, action_input, name=None)`)